src/components/TodoFilter/TodoFilter.tsx

```diff
@@ -0,0 +1,133 @@
+import debounce from 'lodash.debounce';
+import React, { useCallback, useEffect, useState } from 'react';
+import { useLocation, useNavigate } from 'react-router-dom';
+import { useAppSelector } from '../../store';
+import { Todo } from '../../types/Todo';
+
+interface Props {
+  onSetVisibleTodos: React.Dispatch<React.SetStateAction<Todo[]>>
+}
+
+enum Selector {
+  all = 'all',
+  active = 'active',
+  completed = 'completed',
+}
+
+export const TodoFilter: React.FC<Props> = ({ onSetVisibleTodos }) => {
+  const todos = useAppSelector(state => state.todos);
+  const navigation = useNavigate();
+  const location = useLocation();
+  const searchParams = new URLSearchParams(location.search);
+
+  const appliedQuery = searchParams.get('query') || '';
+  const [query, setQuery] = useState(appliedQuery);
+
+  const applyQuery = useCallback(
+    debounce((newQuery: string) => {
+      if (newQuery) {
+        searchParams.set('query', newQuery);
+      } else {
+        searchParams.delete('query');
+      }
+
+      navigation(`?${searchParams}`, { replace: true });
+    }, 500),
+    [location.search],
+  );
+
+  const handleQueryChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setQuery(event.target.value);
+    applyQuery(event.target.value);
+  };
+
+  const appliedSelect = searchParams.get('select') || '';
+  const [selectFilter, setSelectFilter] = useState(appliedSelect);
+
+  const handleSelectChange = ({
+    target,
+  }: React.ChangeEvent<HTMLSelectElement>) => {
+    if (target.value === Selector.all) {
+      searchParams.delete('select');
+    } else {
+      searchParams.set('select', target.value);
+    }
+
+    navigation(`?${searchParams}`, { replace: true });
+    setSelectFilter(target.value);
+  };
+
+  const selectFilterCallback = (todo: Todo) => {
+    const pickFilterParam = (str: string) => {
+      switch (location.search.includes(str)) {
+        case str === '':
+          return todo;
+
+        case str === Selector.active:
+          return !todo.completed;
+
+        case str === Selector.completed:
+          return todo.completed;
+
+        default:
+          return todo;
+      }
```

Comment:
1) looks weird that you use switch-case this way – you compare two booleans, you'd rather compare `Selector` values only
2) in switch-case, you have to return item of only one type, and here you return either `Todo` or boolean – not OK